clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -87,7 +87,7 @@ public FetchResponseData data() {
      * We may also return INCONSISTENT_TOPIC_ID error as a partition-level error when a partition in the session has a topic ID
      * inconsistent with the log.
      */
-    public FetchResponse(FetchResponseData fetchResponseData) {
+    FetchResponse(FetchResponseData fetchResponseData) {
```

Comment:
I thought that we wanted to make this private so that all callers on the server side will go through the `of` constructor that has the null check on `records`?

Reply:
According to https://github.com/apache/kafka/pull/19167#discussion_r2024085572
If we have a private scope for the constructor, the request `version < 13`[0] will have an overhead (iterate 2 times).
If we worry about the wrongly invoked constructor, we could add a Javadoc to it. Although it's not good protection, it seems like a tradeoff for code style and performance.
WDYT?

[0]
https://github.com/apache/kafka/blob/e4792c4b01b3127b0019b5469c51e72b96876e7d/clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java#L353-L362

Reply:
@frankvicky discussed that with me offline, and I agree that is a tradeoff. Also, making it package-private is good enough to obstruct server-side from calling constructor directly.

Reply:
Looking at the relevant commit, it seems like the double iteration is only for the error handling case. That doesn't seem a major issue, if true.

Reply:
> Looking at the relevant commit, it seems like the double iteration is only for the error handling case. That doesn't seem a major issue, if true.

yes, that is not hot path. I'm fine with private constructor :)